client/executor/common/src/sandbox/wasmi_backend.rs

```diff
@@ -0,0 +1,341 @@
+				invoke_args_ptr,
+				"Can't deallocate memory for dispatch thunk's invoke arguments",
+			)?;
+			let result = result?;
+
+			// dispatch_thunk returns pointer to serialized arguments.
+			// Unpack pointer and len of the serialized result data.
+			let (serialized_result_val_ptr, serialized_result_val_len) = {
+				// Cast to u64 to use zero-extension.
+				let v = result as u64;
+				let ptr = (v as u64 >> 32) as u32;
+				let len = (v & 0xFFFFFFFF) as u32;
+				(Pointer::new(ptr), len)
+			};
+
+			let serialized_result_val = sandbox_context
+				.supervisor_context()
+				.read_memory(serialized_result_val_ptr, serialized_result_val_len)
+				.map_err(|_| trap("Can't read the serialized result from dispatch thunk"));
+
+			deallocate(
+				sandbox_context.supervisor_context(),
+				serialized_result_val_ptr,
+				"Can't deallocate memory for dispatch thunk's result",
+			)
+			.and_then(|_| serialized_result_val)
+			.and_then(|serialized_result_val| {
+				let result_val = std::result::Result::<ReturnValue, HostError>::decode(&mut serialized_result_val.as_slice())
+					.map_err(|_| trap("Decoding Result<ReturnValue, HostError> failed!"))?;
+
+				match result_val {
+					Ok(return_value) => Ok(match return_value {
+						ReturnValue::Unit => None,
+						ReturnValue::Value(typed_value) => Some(RuntimeValue::from(typed_value)),
+					}),
+					Err(HostError) => Err(trap("Supervisor function returned sandbox::HostError")),
+				}
+			})
+		}).expect("SandboxContextStore is set when invoking sandboxed functions; qed")
+	}
+}
+
+fn with_guest_externals<R, F>(sandbox_instance: &SandboxInstance, state: u32, f: F) -> R
+where
+	F: FnOnce(&mut GuestExternals) -> R,
+{
+	f(&mut GuestExternals { sandbox_instance, state })
+}
+
+pub fn instantiate(
+	wasm: &[u8],
+	guest_env: GuestEnvironment,
+	state: u32,
+	sandbox_context: &mut dyn SandboxContext,
+) -> std::result::Result<Rc<SandboxInstance>, InstantiationError> {
+	let wasmi_module = Module::from_buffer(wasm).map_err(|_| InstantiationError::ModuleDecoding)?;
+	let wasmi_instance = ModuleInstance::new(&wasmi_module, &guest_env.imports)
+		.map_err(|_| InstantiationError::Instantiation)?;
+
+	let sandbox_instance = Rc::new(SandboxInstance {
+		// In general, it's not a very good idea to use `.not_started_instance()` for
+		// anything but for extracting memory and tables. But in this particular case, we
+		// are extracting for the purpose of running `start` function which should be ok.
+		backend_instance: BackendInstance::Wasmi(wasmi_instance.not_started_instance().clone()),
+		guest_to_supervisor_mapping: guest_env.guest_to_supervisor_mapping,
+	});
+
+	with_guest_externals(&sandbox_instance, state, |guest_externals| {
+		SandboxContextStore::using(sandbox_context, || {
+			wasmi_instance
+				.run_start(guest_externals)
+				.map_err(|_| InstantiationError::StartTrapped)
+		})
+
+		// Note: no need to run start on wasmtime instance, since it's done
```

Comment:
Copy & paste?

Reply:
No, it's a legacy from the times when wasmtime/lightbeam was used as a sandbox backend. Removed.